### PR TITLE
chore: upgrade go toolchain to 1.23

### PR DESCRIPTION
**Pull request by rbtr**
**Reason for Change**:

Updates the Go version to v1.23.1 for various improvements, fixes to stdlib, and toolchain upgrades.

**Issue Fixed**:

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] relevant PR labels added

**Notes**:

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,6 +1,6 @@
 module github.com/Azure/azure-container-networking
 
-go 1.22.0
+go 1.23.1
```
Comment (jpayne3506) [inline go.mod:3]: Setting the required go version to be above that of the defined toolchain depreciates its usage.

https://go.dev/ref/mod#go-mod-file-toolchain

**Review by jpayne3506 — APPROVED** [review]
While it is not a priority. We may want to look at leveraging the toolchain directive to have reproduceable builds for each of our releases/tags.

**Comment by jpayne3506** [comment]
Going to need to re-run the workflows with an updated ruleset for this to merge.

File: go.mod [hunk 2]
```diff
@@ -1,8 +1,6 @@
 module github.com/Azure/azure-container-networking
 
-go 1.22.0
-
-toolchain go1.22.7
+go 1.23
```
Comment (jpayne3506) [inline go.mod:3]: Opinions on setting the toolchain to 1.23.1?

File: .github/workflows/crdgen.yaml [hunk 3]
```diff
@@ -31,5 +31,9 @@ jobs:
       run: make -C crd/multitenantnetworkcontainer
     - name: Regenerate Multitenancy CRDs
       run: make -C crd/multitenancy
+    - name: Regenerate ClusterSubnetState CRD
+      run: make -C crd/clustersubnetstate
+    - name: Regenerate OverlayExtensionConfig CRD
+      run: make -C crd/overlayextensionconfig
```
Comment (jpayne3506) [inline .github/workflows/crdgen.yaml:37]: Is this something new with 1.23? Or should this be its own PR so we can backport it?

Comment (rbtr) [reply .github/workflows/crdgen.yaml:37]: it's fine if these are only in master, we just should have had the CI checking them already

File: go.mod (hunk 2 above)
Comment (rbtr) [reply go.mod:3]: now that the toolchain can upgrade itself based on the `toolchain` directive, how do we get something like dependabot to keep that line up to date?

Comment (rbtr) [reply go.mod:3]: updated to 1.23.2

**Review by aggarwal0009 — APPROVED** [review]

**Review by paulyufan2 — APPROVED** [review]

**Review by huntergregory — DISMISSED** [review]

**Review by jpayne3506 — DISMISSED** [review]

**Review by camrynl — DISMISSED** [review]

**Review by ZetaoZhuang — APPROVED** [review]

**Comment by rbtr** [comment]
this passed all pipeline checks and previously had full approvals, so i'm going to merge it
